pkg/profile/builder/builder.go

```diff
@@ -0,0 +1,349 @@
+package builder
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	helmv2 "github.com/fluxcd/helm-controller/api/v2beta1"
+	kustomizev1 "github.com/fluxcd/kustomize-controller/api/v1beta1"
+	"github.com/fluxcd/pkg/apis/meta"
+	"github.com/fluxcd/pkg/runtime/dependency"
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
+	profilesv1 "github.com/weaveworks/profiles/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+	"sigs.k8s.io/kustomize/api/types"
+
+	"github.com/weaveworks/pctl/pkg/profile/artifact"
+)
+
+const (
+	defaultValuesKey           = "default-values.yaml"
+	helmChartLocation          = "helm-chart"
+	kustomizeWrapperObjectName = "kustomize-flux"
+)
+
+// Builder can build an artifacts from an installation and a profile artifact.
+//go:generate counterfeiter -o fakes/builder_maker.go . Builder
+type Builder interface {
+	// Build a single artifact from a profile artifact and installation.
+	Build(artifact profilesv1.Artifact, installation profilesv1.ProfileInstallation, definition profilesv1.ProfileDefinition) ([]artifact.Artifact, error)
+}
+
+// Config defines some common configuration values for builders.
+type Config struct {
+	GitRepositoryName      string
+	GitRepositoryNamespace string
+	RootDir                string
+}
+
+// ArtifactBuilder will build helm chart resources.
+type ArtifactBuilder struct {
+	Config
+	a *artifact.Artifact
+}
+
+// Build a single artifact from a profile artifact and installation.
+func (c *ArtifactBuilder) Build(partifact profilesv1.Artifact, installation profilesv1.ProfileInstallation, definition profilesv1.ProfileDefinition) ([]artifact.Artifact, error) {
```

Comment:
feels weird we pass in the artifact and the whole definition, but I guess its unavoidable for this check
```
	for _, dep := range partifact.DependsOn {
		d, ok := c.containsArtifact(dep.Name, definition.Spec.Artifacts)
		if !ok {
			return nil, fmt.Errorf("%s's depending artifact %s not found in the list of artifacts", partifact.Name, dep.Name)
		}
		deps = append(deps, d)
	}
```

Reply:
perhaps we could move this logic outside of the builder?

Reply:
I think I can perform it in the Make phase. I'll take a look.

Reply:
ah! I know why I'm doing this here. :) I'm doing this here, because we need the dependencies and checking them outside, in Make, I would still have to pass them in, which would mean, that I have to modify the Build interface, which don't want. And as you saw, since this has to be performed for each artifact, and various other things, I have to also have the definition. :/